### PR TITLE
Add adhoc and playbook subcommands for CLI Exec Env usage

### PR DESCRIPTION
**Pull request by maxamillion**
Currently with Execution Environments via process isolation option, the interface to that containerized functionality is either via the native Ansible Runner CLI or the corresponding python API. In an attempt to open this functionality up in a more "familiar feel" to the users of the `ansible` and `ansible-playbook` command line tools, in service of aiding in the dev/test of automation targeted for Tower/AWS, I wanted to introduce the `adhoc` and `playbook` subcommands to the `ansible-runner` CLI (while maintaining backward compatibility with the Runner CLI that exists today). The main goal here is to allow an user to have a familiar feel and even potentially use a shell alias that would allow them to use these subcommands the same way they use `ansible` and `ansible-runner` today.

I've provided the new CLI functionality that targets both `podman` and `docker` (along with the help of @shanemcd, @kdelee, and @jladdjr) as compatible OCI compliant runtimes, I've also included some documentation, and a couple baseline test cases. There's is likely room for improvement on all of these areas but I wanted to go ahead and get the PR open so we can start the review/feedback/iterate process.

Examples using my topic branch and `poetry` to venv management:
`ansible`
```
$ poetry run ansible-runner adhoc localhost -m ping
[WARNING]: No inventory was parsed, only implicit localhost is available
localhost | SUCCESS => {
    "changed": false,
    "ping": "pong"
}

$ alias ansible="poetry run ansible-runner adhoc"

$ which ansible
alias ansible='poetry run ansible-runner adhoc'
        ~/.local/bin/poetry

$ ansible localhost -m ping
[WARNING]: No inventory was parsed, only implicit localhost is available
localhost | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
```
`ansible-playbook`
```
$ poetry run ansible-runner playbook test/integration/exec_env/demo.yml -i test/integration/exec_env/inventory.ini

PLAY [Run some basic system automation stuff] **********************************

TASK [Gathering Facts] *********************************************************
ok: [azure-rhel8]

TASK [install some basic packages] *********************************************
ok: [azure-rhel8]

TASK [enable journald persistent storage] **************************************
ok: [azure-rhel8]

TASK [Get tuned profile] *******************************************************
ok: [azure-rhel8]

TASK [debug] *******************************************************************
ok: [azure-rhel8] => {
    "msg": "throughput-performance"
}

TASK [tuned-adm set throughput-performance] ************************************
skipping: [azure-rhel8]

TASK [don't allow password based ssh] ******************************************
ok: [azure-rhel8]

PLAY RECAP *********************************************************************
azure-rhel8                : ok=6    changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0

$ alias ansible-playbook="poetry run ansible-runner playbook"

$ which ansible-playbook
alias ansible-playbook='poetry run ansible-runner playbook'
        ~/.local/bin/poetry

$ ansible-playbook test/integration/exec_env/demo.yml -i test/integration/exec_env/inventory.ini

PLAY [Run some basic system automation stuff] **********************************

TASK [Gathering Facts] *********************************************************
ok: [azure-rhel8]

TASK [install some basic packages] *********************************************
ok: [azure-rhel8]

TASK [enable journald persistent storage] **************************************
ok: [azure-rhel8]

TASK [Get tuned profile] *******************************************************
ok: [azure-rhel8]

TASK [debug] *******************************************************************
ok: [azure-rhel8] => {
    "msg": "throughput-performance"
}

TASK [tuned-adm set throughput-performance] ************************************
skipping: [azure-rhel8]

TASK [don't allow password based ssh] ******************************************
ok: [azure-rhel8]

PLAY RECAP *********************************************************************
azure-rhel8                : ok=6    changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
```

Thank you,
-AdamM

### CODE REVIEW

**Comment by shanemcd** [comment]
I'm playing around with this now. I'm going to post tracebacks as separate comments as I find them.

**Comment by shanemcd** [comment]
Hitting this in multiple scenarios:

```
$ poetry run ansible-runner adhoc -m debug -a 'msg=hi' localhost
Traceback (most recent call last):
  File "/home/shanemcd/ansible-runner/ansible_runner/__main__.py", line 945, in main
    res = run(**run_options)
  File "/home/shanemcd/ansible-runner/ansible_runner/interface.py", line 210, in run
    r = init_runner(**kwargs)
  File "/home/shanemcd/ansible-runner/ansible_runner/interface.py", line 75, in init_runner
    rc.prepare()
  File "/home/shanemcd/ansible-runner/ansible_runner/runner_config.py", line 274, in prepare
    self.command = self.wrap_args_for_containerization(self.command)
  File "/home/shanemcd/ansible-runner/ansible_runner/runner_config.py", line 654, in wrap_args_for_containerization
    _parsed_playbook_path, _parsed_inventory_paths = _parse_cli_execenv_cmd_playbook_args()
  File "/home/shanemcd/ansible-runner/ansible_runner/runner_config.py", line 647, in _parse_cli_execenv_cmd_playbook_args
    elif _book_keeping_copy[_book_keeping_copy.index(arg - 1)][0] != '-':
TypeError: unsupported operand type(s) for -: 'str' and 'int'
```

Seen with the command above and:

```
poetry run ansible-runner adhoc -i 'example.ansible.com,' -m debug -a 'var=vars' all
```


```
poetry run ansible-runner playbook -i 'example.ansible.com,' -u smcdonald --become demo/project/test.yml
```

**Comment by maxamillion** [comment]
@shanemcd thanks, I'll dig into that one and see what's going on.

File: ansible_runner/runner_config.py [hunk 1]
```diff
@@ -568,21 +611,125 @@ def _ensure_path_safe_to_mount(path):
...
+                        else:
+                            new_args.extend([
+                                "-v", "{}:/runner/project/{}".format(
+                                   os.path.dirname(os.path.abspath(inventory_file_path)),
+                                   os.path.dirname(inventory_file_path),
+                                )
+                            ])
+
+            # volume mount ~/.ssh/ into the exec env container
+            new_args.extend(["-v", "{}/.ssh/:/runner/.ssh/".format(os.environ['HOME'])])
```
Comment (shanemcd) [inline ansible_runner/runner_config.py]: Just noticed I broke `known_hosts` support when I changed the home dir location. This should fix it:

```suggestion
            new_args.extend(["-v", "{}/.ssh/:/home/runner/.ssh/".format(os.environ['HOME'])])
```

File: Dockerfile [hunk 2]
```diff
@@ -10,25 +9,36 @@ ADD demo/inventory /runner/inventory
...
+ADD utils/entrypoint.sh /bin/entrypoint
+RUN chmod +x /bin/entrypoint
+
 # In OpenShift, container will run as a random uid number and gid 0. Make sure things
 # are writeable by the root group.
-RUN mkdir -p /runner/inventory /runner/project /runner/artifacts /runner/.ansible/tmp && \
-	chmod -R g+w /runner && chgrp -R root /runner && \
-	chmod g+w /etc/passwd
+RUN for dir in \
+      /runner \
```
Comment (shanemcd) [inline Dockerfile:31]: Ran into some permissions errors. This fixes it:

```suggestion
      /runner \
      /home/runner \
```

**Comment by maxamillion** [comment]
> Hitting this in multiple scenarios:
> 
> ```
> $ poetry run ansible-runner adhoc -m debug -a 'msg=hi' localhost
> Traceback (most recent call last):
>   File "/home/shanemcd/ansible-runner/ansible_runner/__main__.py", line 945, in main
>     res = run(**run_options)
>   File "/home/shanemcd/ansible-runner/ansible_runner/interface.py", line 210, in run
>     r = init_runner(**kwargs)
>   File "/home/shanemcd/ansible-runner/ansible_runner/interface.py", line 75, in init_runner
>     rc.prepare()
>   File "/home/shanemcd/ansible-runner/ansible_runner/runner_config.py", line 274, in prepare
>     self.command = self.wrap_args_for_containerization(self.command)
>   File "/home/shanemcd/ansible-runner/ansible_runner/runner_config.py", line 654, in wrap_args_for_containerization
>     _parsed_playbook_path, _parsed_inventory_paths = _parse_cli_execenv_cmd_playbook_args()
>   File "/home/shanemcd/ansible-runner/ansible_runner/runner_config.py", line 647, in _parse_cli_execenv_cmd_playbook_args
>     elif _book_keeping_copy[_book_keeping_copy.index(arg - 1)][0] != '-':
> TypeError: unsupported operand type(s) for -: 'str' and 'int'
> ```
> 
> Seen with the command above and:
> 
> ```
> poetry run ansible-runner adhoc -i 'example.ansible.com,' -m debug -a 'var=vars' all
> ```
> 
> ```
> poetry run ansible-runner playbook -i 'example.ansible.com,' -u smcdonald --become demo/project/test.yml
> ```

@shanemcd I've fixed this, I was trying to perform an arithmetic operation on an index but didn't quote it correctly and inadvertently ended up attempting the arithmetic operation on a str type. Fun times, thank you for pointing that out. :)

**Comment by shanemcd** [comment]
Here's a new one for you:

```
$ poetry run ansible-runner run --process-isolation /root/ansible-runner/demo --playbook /root/ansible-runner/demo/project/test.yml
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/shanemcd/ansible-runner/ansible_runner/__main__.py", line 818, in main
    vargs = vars(args)
TypeError: vars() argument must have __dict__ attribute
```

**Comment by shanemcd** [comment]
Currently trying to figure this one out:

```
$ poetry run ansible-runner run -p test.yml demo
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/shanemcd/ansible-runner/ansible_runner/__main__.py", line 818, in main
    vargs = vars(args)
TypeError: vars() argument must have __dict__ attribute
```

You can see here that `args` is a `argparse.Namespace`, wrapped in a tuple:

```
(Pdb) print(args)
(Namespace(artifact_dir=None, binary=None, cmdline=None, command='run', container_image='ansible/ansible-runner', container_options=None, container_volume_mounts=None, debug=False, directory_isolation_base_path=None, forks=None, hosts=None, ident=UUID('b0c98cc1-c37d-4426-91ce-553c541cd96b'), inventory=None, json=False, limit=None, logfile=None, module=None, module_args=None, omit_event_data=False, only_failed_event_data=False, playbook='test.yml', private_data_dir='demo', process_isolation=False, process_isolation_executable='podman', process_isolation_hide_paths=None, process_isolation_path='/tmp', process_isolation_ro_paths=None, process_isolation_show_paths=None, project_dir=None, quiet=False, receptor_node_id=None, receptor_peer=None, resource_profiling=False, resource_profiling_base_cgroup='ansible-runner', resource_profiling_cpu_poll_interval=0.25, resource_profiling_memory_poll_interval=0.25, resource_profiling_pid_poll_interval=0.25, resource_profiling_results_dir=None, role=None, role_skip_facts=False, role_vars=None, roles_path=None, rotate_artifacts=0, v=None, via_receptor=None), [])
(Pdb) type(args)
<class 'tuple'>
```

But with your new playbook / adhoc subcommands, `args` is just a `argparse.Namespace`:

```
$ poetry run ansible-runner adhoc -m setup
```

```
(Pdb) print(args)
Namespace(artifact_dir=None, binary=None, cmdline=None, command='adhoc', container_image='ansible/ansible-runner', container_options=None, container_runtime='podman', container_volume_mounts=None, debug=False, directory_isolation_base_path=None, forks=None, hosts=None, ident=UUID('8cb99c8b-4eb7-408c-9b07-4000c4139064'), inventory=None, json=False, keep_files=False, limit=None, logfile=None, module=None, module_args=None, omit_event_data=False, only_failed_event_data=False, playbook=None, private_data_dir=None, process_isolation=False, process_isolation_executable='podman', process_isolation_hide_paths=None, process_isolation_path='/tmp', process_isolation_ro_paths=None, process_isolation_show_paths=None, project_dir=None, quiet=False, receptor_node_id=None, receptor_peer=None, resource_profiling=False, resource_profiling_base_cgroup='ansible-runner', resource_profiling_cpu_poll_interval=0.25, resource_profiling_memory_poll_interval=0.25, resource_profiling_pid_poll_interval=0.25, resource_profiling_results_dir=None, role=None, role_skip_facts=False, role_vars=None, roles_path=None, rotate_artifacts=0, v=None, via_receptor=None)
(Pdb) type(args)
<class 'argparse.Namespace'>
```

File: Dockerfile [hunk 3]
```diff
@@ -1,7 +1,6 @@
-FROM centos:8
+FROM docker.io/centos:8
```
Comment (matburt) [inline Dockerfile:1]: why is this necessary?

File: Dockerfile [hunk 4]
```diff
@@ -10,25 +9,40 @@ ADD demo/inventory /runner/inventory
...
+    do touch $file ; chmod g+rw $file ; chgrp runner $file ; done
 
 VOLUME /runner
 
 ENV LANG=en_US.UTF-8
 ENV LANGUAGE=en_US:en
 ENV LC_ALL=en_US.UTF-8
 ENV RUNNER_BASE_COMMAND=ansible-playbook
-ENV HOME=/runner
+ENV HOME=/home/runner
```
Comment (matburt) [inline Dockerfile:49]: This is a pretty major change to the overall directory structure of the image and is going to cause problems for folks like the operator. For better or worse there are some assumptions that are baked in to `/runner` being HOME.

File: ansible_runner/runner_config.py [hunk 5]
```diff
@@ -53,6 +53,7 @@ class ExecutionMode():
     ANSIBLE = 1
     ANSIBLE_PLAYBOOK = 2
     RAW = 3
+    CLI_EXECENV = 4
```
Comment (matburt) [inline ansible_runner/runner_config.py:56]: I might need help understanding why this needs another mode rather than just modifying the overall command.

File: ansible_runner/runner_config.py [hunk 6]
```diff
@@ -380,22 +383,41 @@ def prepare_command(self):
...
         """
-        try:
-            cmdline_args = self.loader.load_file('args', string_types, encoding=None)
-            if six.PY2 and isinstance(cmdline_args, text_type):
-                cmdline_args = cmdline_args.encode('utf-8')
-            self.command = shlex.split(cmdline_args)
-            self.execution_mode = ExecutionMode.RAW
-        except ConfigurationError:
-            self.command = self.generate_ansible_command()
+        if not self.cli_execenv_cmd:
```
Comment (matburt) [inline ansible_runner/runner_config.py:386]: I can kinda see where you are going with this in that you want to pivot the mode around looking like ansible execution. I think I'd rather we do some refactoring around command-line generation, though.....

File: ansible_runner/runner_config.py [hunk 7]
```diff
@@ -407,6 +429,27 @@ def generate_ansible_command(self):
 
         exec_list = [base_command]
 
+        if self.cli_execenv_cmd:
```
Comment (matburt) [inline ansible_runner/runner_config.py:432]: Why would they fail? We haven't needed these before.

File: ansible_runner/runner_config.py [hunk 8]
```diff
@@ -563,26 +606,136 @@ def wrap_args_for_containerization(self, args):
...
                 raise ConfigurationError("When using containerized execution, cannot mount /home or /usr")
 
-        # map private data to /runner
-        # .. using Z to modify the selinux label of the _host_ file / directory
-        # (see https://docs.docker.com/storage/bind-mounts/#configure-the-selinux-label
-        #  for usage and potential side-effects)
         _ensure_path_safe_to_mount(self.private_data_dir)
-        new_args.extend(["-v", "{}:/runner:Z".format(self.private_data_dir)])
+
+        if not self.cli_execenv_cmd:
```
Comment (matburt) [inline ansible_runner/runner_config.py]: I'm particularly worried about the overall state machine for command line generation, the complexity is a bit out of control and I'd like to change this to a more composable system.

File: Dockerfile (hunk 3 above)
Comment (shanemcd) [reply Dockerfile:1]: Podman on RHEL 8 doesn't have docker configured as a default registry, and there is no centos image on registry.redhat.io.

File: Dockerfile (hunk 4 above)
Comment (shanemcd) [reply Dockerfile:49]: @matburt This isn't changing where the host-dirs are mounted, only changing $HOME. What are the specific problems you think this will cause? There were some really gnarly things going on here. I will revert this part and post the error.

**Comment by shanemcd** [comment]
One of my commits yesterday broke something else 😅 :

```
$ poetry run ansible-runner --version
usage: ansible-runner [-h] [-b BINARY] [-i IDENT]
                      [--rotate-artifacts ROTATE_ARTIFACTS]
                      [--artifact-dir ARTIFACT_DIR]
                      [--project-dir PROJECT_DIR] [--inventory INVENTORY] [-j]
                      [--omit-event-data] [--only-failed-event-data] [-q] [-v]
                      [--via-receptor VIA_RECEPTOR]
                      [--receptor-peer RECEPTOR_PEER]
                      [--receptor-node-id RECEPTOR_NODE_ID]
                      [-p PLAYBOOK | -m MODULE | -r ROLE] [--limit LIMIT]
                      [--cmdline CMDLINE] [--hosts HOSTS] [--forks FORKS]
                      [--roles-path ROLES_PATH] [--role-vars ROLE_VARS]
                      [--role-skip-facts] [-a MODULE_ARGS]
                      [--process-isolation]
                      [--process-isolation-executable PROCESS_ISOLATION_EXECUTABLE]
                      [--process-isolation-path PROCESS_ISOLATION_PATH]
                      [--process-isolation-hide-paths [PROCESS_ISOLATION_HIDE_PATHS [PROCESS_ISOLATION_HIDE_PATHS ...]]]
                      [--process-isolation-show-paths [PROCESS_ISOLATION_SHOW_PATHS [PROCESS_ISOLATION_SHOW_PATHS ...]]]
                      [--process-isolation-ro-paths [PROCESS_ISOLATION_RO_PATHS [PROCESS_ISOLATION_RO_PATHS ...]]]
                      [--directory-isolation-base-path DIRECTORY_ISOLATION_BASE_PATH]
                      [--resource-profiling]
                      [--resource-profiling-base-cgroup RESOURCE_PROFILING_BASE_CGROUP]
                      [--resource-profiling-cpu-poll-interval RESOURCE_PROFILING_CPU_POLL_INTERVAL]
                      [--resource-profiling-memory-poll-interval RESOURCE_PROFILING_MEMORY_POLL_INTERVAL]
                      [--resource-profiling-pid-poll-interval RESOURCE_PROFILING_PID_POLL_INTERVAL]
                      [--resource-profiling-results-dir RESOURCE_PROFILING_RESULTS_DIR]
                      [--container-image CONTAINER_IMAGE]
                      [--container-volume-mount CONTAINER_VOLUME_MOUNTS]
                      [--container-option CONTAINER_OPTIONS]
                      {run,start,stop,is-alive,adhoc,playbook} ...
ansible-runner: error: the following arguments are required: command
```

**Review by shanemcd — APPROVED** [review]

**Comment by AlanCoding** [comment]
I'm having trouble with these tests and would like to sync up with someone about the right amount of effort to patch them up. 

Testing locally, it seems that they are all running the bare command of `ansible-runner` which returns the help text and gives an exit code of 0. Since no other assertions are made, all the tests just think they passed. This may not be true in other environments.. but it probably is. I want to figure out exactly where these are expected to run, because they get skipped in Zuul. I plan to add more tests, but not if it's not actually working.

**Comment by jladdjr** [comment]
Earlier today, I noticed that the tests failed for me when I didn't have a local image for `ansible/ansible-runner`. Those tests passed for me once I built the image. I'm wondering if this change makes local images necessary:

https://github.com/ansible/ansible-runner/commit/086011b4#diff-c5c59c296282fc1a82b191b56a6d65ddR31

I built the runner image with:
`podman build -t ansible/ansible-runner -f Dockerfile .`

**Comment by AlanCoding** [comment]
I put up #509 after consolidating my findings from this

File: ansible_runner/runner_config.py (hunk 7 above)
Comment (AlanCoding) [reply ansible_runner/runner_config.py:432]: Even accepting the need for the hack, the way this is implemented concerns me, because these will be defined in the extravars file, but `-e` with dummy data could still get added in conflict with those.
